firebase-firestore/src/test/java/com/google/firebase/firestore/model/MutationSquashTest.java

```diff
@@ -0,0 +1,250 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+/** Tests Mutations */
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = Config.NONE)
+public class MutationSquashTest {
+  @Test
+  public void testSquashOneSetMutation() {
+    Map<String, Object> data = map("foo", "foo-value", "baz", "baz-value");
+    MutableDocument setDoc = doc("collection/key", 0, data);
+    MutableDocument original = setDoc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation set = setMutation("collection/key", map("bar", "bar-value"));
+    set.applyToLocalView(setDoc, now, MutationSquash.Type.None);
+
+    Mutation squashed = MutationSquash.toMutation(original, setDoc, MutationSquash.Type.Set);
+    squashed.applyToLocalView(original, now, MutationSquash.Type.None);
+    assertEquals(original, setDoc);
+  }
+
+  @Test
+  public void testSquashOnePatchMutation() {
+    Map<String, Object> data = map("foo", map("bar", "bar-value"), "baz", "baz-value");
+    MutableDocument patchDoc = doc("collection/key", 0, data);
+    MutableDocument original = patchDoc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation patch = patchMutation("collection/key", map("foo.bar", "new-bar-value"));
+    patch.applyToLocalView(patchDoc, now, MutationSquash.Type.None);
+
+    Mutation squashed = MutationSquash.toMutation(original, patchDoc, MutationSquash.Type.Patch);
+    squashed.applyToLocalView(original, now, MutationSquash.Type.None);
+    assertEquals(original, patchDoc);
+  }
+
+  @Test
+  public void testSquashWithPatchWithMerge() {
+    MutableDocument mergeDoc = deletedDoc("collection/key", 0);
+    MutableDocument original = mergeDoc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation upsert =
+        mergeMutation(
+            "collection/key", map("foo.bar", "new-bar-value"), Arrays.asList(field("foo.bar")));
+    MutationSquash.Type type = upsert.applyToLocalView(mergeDoc, now, MutationSquash.Type.None);
+
+    Mutation squashed = MutationSquash.toMutation(original, mergeDoc, type);
+    squashed.applyToLocalView(original, now, MutationSquash.Type.None);
+    assertEquals(original, mergeDoc);
+  }
+
+  @Test
+  public void testSquashWithDeleteThenPatch() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    MutableDocument original = doc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation delete = new DeleteMutation(key("collection/key"), Precondition.NONE);
+    delete.applyToLocalView(doc, now, MutationSquash.Type.None);
+
+    now = Timestamp.now();
+    Mutation patch = patchMutation("collection/key", map("foo.bar", "new-bar-value"));
+    patch.applyToLocalView(doc, now, MutationSquash.Type.None);
+
+    Mutation squashed = MutationSquash.toMutation(original, doc, MutationSquash.Type.Delete);
+    squashed.applyToLocalView(original, now, MutationSquash.Type.None);
+    assertEquals(doc, original);
```

Comment:
`original` should go first - but I think you can just encode the final state here. It's a deleted doc, right? Your test hides this logic.